### ISSUE TITLE
XML glyph renders with an extra space before filename

### ISSUE THREAD

**Issue by tillig** [devblackops/Terminal-Icons#34]
## Expected Behavior

Most of the icons render in a single character width; I expect the XML icon to also render in one character.

## Current Behavior

On both Windows (PowerShell 7.1.3 / Windows Terminal) and Mac (PowerShell 7.1.3 / iTerm2) the XML icons appear to insert a leading space before the filename. Other icons don't do that.

Here's JetBrainsMono Nerd Font:

<img width="574" alt="JetBrainsMono Nerd Font" src="https://user-images.githubusercontent.com/1156571/115308934-66c2cd00-a120-11eb-827e-108db004c138.png">

Here's FiraCode Nerd Font:

<img width="653" alt="FiraCode Nerd Font" src="https://user-images.githubusercontent.com/1156571/115308996-7f32e780-a120-11eb-92bd-5e91ffdf2119.png">

It doesn't seem to matter about the font or the terminal, there's always that extra space.

## Possible Solution

I'm not sure if it's a double-character glyph where the extra space is an artifact or if there's a hidden character in the icon theme I haven't seen yet. Double-wide glyphs can get weird.

## Steps to Reproduce (for bugs)

1. Set up your terminal to use FiraCode Nerd Font or JetBrainsMono Nerd Font. Doesn't seem to matter if it's Mac or Windows.
2. List a folder with an XML file in it.

## Context

New user of the module, just noticed the odd spacing when listing a folder.

## Your Environment

* Module version used: 0.4.0
* Operating System and PowerShell version: MacOS Big Sur 11.2.3, PowerShell 7.1.3, iTerm2

**Comment by tillig** [comment]
Definitely some weird rendering thing going on with the double-wide glyph. If I copy/paste a listing into VS Code, I can see that the file listing lines up, but you can just make out that the XML glyph is covering one of the two spaces.

<img width="899" alt="Listing in VS Code Editor" src="https://user-images.githubusercontent.com/1156571/115313219-b9ec4e00-a127-11eb-823a-b73640c1862e.png">

Same font and everything, but in the Terminal window of VS Code shows the glyph "pushes" the space over and throws everything off.

<img width="904" alt="VS Code Terminal Window" src="https://user-images.githubusercontent.com/1156571/115313307-e0aa8480-a127-11eb-9517-0bca17e300f9.png">

Unclear why it doesn't have the same effect on the directories, which also appear to be double-wide.

I'm guessing the short-term resolution is to pick a different icon for XML files that's not double-wide and just call it good.

**Comment by tillig** [comment]
Here's a snippet to replace the double-wide XML icon with something single-width. It doesn't appear I can access the actual `glyphs.ps1` data so I can't replace `nf-mdi-xml` with something different at the top level; instead I need to update all the mappings. Not a huge deal, but it might be nice to have a "swap glyph x with glyph y" command for stuff like this.

```powershell
$fileIcons = (Get-TerminalIconsTheme).Icon.Types.Files
$xmlKeys = $fileIcons.Keys | Where-Object { $fileIcons[$_] -eq "nf-mdi-xml" }
$xmlKeys | %{ $fileIcons[$_] = "nf-mdi-file_xml" }
```

**Comment by ShrykeWindgrace** [comment]
In addition to `*.xml` files I am observing the same issue with `*.log` files. Tested on "UbuntuMono Nerd Font" and "InconsolataLGCMono Nerd Font".

**Comment by devblackops** [comment]
Would there be an straightforward way to detect if it's a double-wide glyph? When we write out the name, we add two spaces between the icon and file/folder name. If we can detect a double-wide glyph we can use only one space to things line up.

https://github.com/devblackops/Terminal-Icons/blob/master/Terminal-Icons/Public/Format-TerminalIcons.ps1#L37

**Comment by tillig** [comment]
If it's consistent across terminals (which it seems to be) it might be something that could be generated sitting a build. For example, could you write out a known string, like `[$glyph]` and then do something with the host `RawUI` (or similar) to read where the brackets are? I haven't tried it, but since it's a rendering thing it seems like you'd have to interact with a host implementation.

**Comment by devblackops** [comment]
I think the simplest solution right now is to not use problematic glyphs so I've switch them out in the theme in https://github.com/devblackops/Terminal-Icons/commit/eb36e628b5429ae0cd14f24d7c5bd24d45072d29.

**Comment by devblackops** [comment]
v0.5.0 has been [released](https://www.powershellgallery.com/packages/Terminal-Icons/0.5.0) with these changes.

**Comment by tillig** [comment]
Sweet, thanks!

**Comment by tillig** [comment]
Here's a script that can be used to calculate the glyph widths.

```powershell
<#
This script writes all the glyphs to the console and inspects the console buffer
to see if the glyph takes up one or two characters. This works on Windows
Terminal in PowerShell Core but does not work on MacOS + iTerm2 + PowerShell
Core.

The buffer for a single-width glyph looks like:


'' Gray Black Complete ' ' Gray Black Complete

A double-wide, on the other hand, looks like:

謹
'謹' Gray Black Complete '謹' Gray Black Complete

So, basically, we only have to write the glyph and grab two buffer cells to
figure out the width.
#>

$currentPos = $host.UI.RawUI.CursorPosition.Y
$rec = New-Object System.Management.Automation.Host.Rectangle(0, ($currentPos - 1), 1, ($currentPos - 1))

$glyphs = Get-TerminalIconsGlyphs

$widths = @{}
$glyphs.Keys | ForEach-Object {
    $name = $_
    $glyph = $glyphs[$_]
    Write-Host $glyph
    $buffer = $host.UI.RawUI.GetBufferContents($rec)
    $widths[$name] = ($buffer[0, 1].Character -eq $glyph) ? 2 : 1
}

$widths
```

Running that...

```powershell
$widths = ./Get-GlyphWidths.ps1
$widths.Keys | Where-Object { $widths[$_] -eq 2 }
```

...we actually find there are a LOT of double-width glpyhs - 513 if you pipe it through `Measure-Object`.

It may be worth adding a step to the build to generate the hash table (or doing it manually on a periodic basis if the host doesn't support it in GitHub Actions) of glyph-to-width and then using that original idea - write the associated number of spaces after the glyph.

**Comment by devblackops** [comment]
@tillig I'm not sure why, but when I run this all the glyphs show as `1` width, yet when I run the below I can clearly see the double-wide glyphs:

```powershell
(Get-TerminalIconsGlyphs).Values | % { "$_ test" }
```

**Comment by tillig** [comment]
I tried a couple of things and I think I figured out the difference, at least on Windows:

Under **Fira Code Nerd Font** all the widths _do_ show up as one. Under **Fira Code Nerd Font _Mono_** I see the variations and it correctly reports as two characters instead of one.

So the difference is in the "Mono" font vs. the one not marked "Mono" - the buffer reports differently based on the font used, even if the final rendering looks the same.

**Comment by devblackops** [comment]
Yep, I see that with FiraCode NF. I usually use SauceCodePro myself.
